client.go

```diff
@@ -101,20 +83,22 @@ func makeHTTPSURL(u *url.URL, addr string) (httpsURL url.URL) {
 
 func (f *federationTripper) RoundTrip(r *http.Request) (*http.Response, error) {
 	serverName := ServerName(r.URL.Host)
-	dnsResult, err := LookupServer(serverName)
+	resolutionResults, err := ResolveServer(serverName)
 	if err != nil {
 		return nil, err
 	}
 
-	if len(dnsResult.Addrs) == 0 {
+	if len(resolutionResults) == 0 {
 		return nil, fmt.Errorf("no address found for matrix host %v", serverName)
 	}
 
 	var resp *http.Response
 	// TODO: respect the priority and weight fields from the SRV record
-	for _, addr := range dnsResult.Addrs {
-		u := makeHTTPSURL(r.URL, addr)
+	for _, result := range resolutionResults {
+		u := makeHTTPSURL(r.URL, result.Destination)
 		r.URL = &u
+		r.Host = string(result.Host)
+		f.transport.(*http.Transport).TLSClientConfig.ServerName = result.Name
```

Comment:
Is it not the case that a given federationTransport can be used for multiple concurrent outbound requests, so this could end up interfering with other requests which happen at the same time?

Reply:
Yes, but as far as I'm aware that's currently the only way to set the server name.

Reply:
I see what you mean :/

I think it's a bit of a show-stopper though, and that we need to find a solution to it. It might be that we need a separate `http.Transport` for each server name?

Reply:
That sound like a decent solution.